Boston Linear Regression Assignments. (Ridge and Lasso Model.) Reference:http://scikit-learn.org/stable/auto_examples/linear_model/plot_lasso_and_elasticnet.html

In [46]:
from sklearn import datasets
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cross_validation import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression


In [47]:
bean = datasets.load_boston()
print bean.DESCR

Boston House Prices dataset

Notes
------
Data Set Characteristics:  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive
    
    :Median Value (attribute 14) is usually the target

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pupil-teacher ratio by town
      

In [3]:
def load_boston():
    scaler = StandardScaler()
    boston = datasets.load_boston()
    X=boston.data
    y=boston.target
    X = scaler.fit_transform(X)
    return train_test_split(X,y)

In [4]:
X_train, X_test, y_train, y_test = load_boston()



In [5]:
X_train.shape



(379, 13)

In [6]:
lr1= LinearRegression()
lr1.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [7]:
zip (y_test, lr1.predict(X_test))

[(11.300000000000001, 13.410164877208654),
 (27.0, 32.808846464516655),
 (19.600000000000001, 19.057894550965937),
 (19.100000000000001, 19.485034619227612),
 (24.800000000000001, 25.952321893089877),
 (22.800000000000001, 28.381382801435983),
 (50.0, 40.149219753806001),
 (21.899999999999999, 37.029366296967012),
 (14.9, 16.968777153641369),
 (50.0, 36.031026904472178),
 (8.3000000000000007, 9.7825595688074358),
 (34.700000000000003, 30.08546239613672),
 (16.600000000000001, 17.759550360467802),
 (25.100000000000001, 30.268955545152387),
 (15.199999999999999, 11.47315547124856),
 (21.5, 20.424692481929192),
 (20.0, 17.393260064114578),
 (23.199999999999999, 17.148462630671951),
 (13.800000000000001, 11.452253209389013),
 (24.0, 25.17660603771769),
 (18.300000000000001, 19.033358581968447),
 (23.199999999999999, 22.201868160119375),
 (24.600000000000001, 28.995429821059684),
 (11.9, 8.5598166174383774),
 (28.600000000000001, 27.96336770954866),
 (24.800000000000001, 30.403135062775654)

In [8]:
from sklearn.linear_model import Lasso

In [15]:
from sklearn.linear_model import Ridge 

In [16]:
from sklearn import linear_model

In [17]:
y_lr1_prediction=lr1.predict(X_test)

In [18]:
r2Score=r2_score(y_test,y_lr1_prediction)

In [19]:
r2Score

0.76410318162264357

In [22]:
msqer_value=mean_squared_error(y_test,y_lr1_prediction)

In [23]:
msqer_value

23.001710078288394

Ridge Model Implementation :

Here alpha value is 0.001

In [25]:
alpha=0.001

Initialisation of Ridge Object 

In [26]:
r_obj=Ridge(alpha=alpha)

fit function fits the training data to the object created for Ridge Linear Regression model

In [27]:
r_obj.fit(X_train,y_train)

Ridge(alpha=0.001, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

Comparison of the real values and the predicted values

In [29]:
zip(y_test,r_obj.predict(X_test))

[(11.300000000000001, 13.410146030023663),
 (27.0, 32.808812638012412),
 (19.600000000000001, 19.057934870487788),
 (19.100000000000001, 19.485004910903072),
 (24.800000000000001, 25.952310965221919),
 (22.800000000000001, 28.381369111430011),
 (50.0, 40.149211573663933),
 (21.899999999999999, 37.029390053498631),
 (14.9, 16.968784583319525),
 (50.0, 36.030998643174939),
 (8.3000000000000007, 9.7825604996094242),
 (34.700000000000003, 30.085469579809448),
 (16.600000000000001, 17.759556361177005),
 (25.100000000000001, 30.268940530232374),
 (15.199999999999999, 11.473291808599077),
 (21.5, 20.424747039744727),
 (20.0, 17.393253910680546),
 (23.199999999999999, 17.148453350447852),
 (13.800000000000001, 11.452341405001725),
 (24.0, 25.17656555863601),
 (18.300000000000001, 19.03339040001503),
 (23.199999999999999, 22.201829690370321),
 (24.600000000000001, 28.995403274799617),
 (11.9, 8.5597628546597377),
 (28.600000000000001, 27.963354602635064),
 (24.800000000000001, 30.40313491346638

In [31]:
y_r_pred=r_obj.predict(X_test)

In [32]:
r2_r_scr=r2_score(y_test,y_r_pred)

In [33]:
r2_r_scr

0.76410331627416939

In [34]:
msqer_r_scr=mean_squared_error(y_test,y_r_pred)

In [35]:
msqer_r_scr


23.001696948754137

R^2 Value for Ridge :0.76410331627416939
MSE Value for Rigde :23.001696948754137


Try at lasso model 


In [36]:
alpha=0.1

In [37]:
l_obj=Lasso(alpha=alpha)

In [38]:
l_obj.fit(X_train,y_train)

Lasso(alpha=0.1, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)

In [39]:
zip(y_test,l_obj.predict(X_test))

[(11.300000000000001, 12.848648081428212),
 (27.0, 32.680982425033115),
 (19.600000000000001, 20.087154856523981),
 (19.100000000000001, 19.133026825546409),
 (24.800000000000001, 25.742620512239831),
 (22.800000000000001, 28.508088359906644),
 (50.0, 40.437159942078445),
 (21.899999999999999, 37.147095211036628),
 (14.9, 17.112870306971658),
 (50.0, 36.338843462828549),
 (8.3000000000000007, 9.8874682059837546),
 (34.700000000000003, 30.481758019984767),
 (16.600000000000001, 17.826969468664405),
 (25.100000000000001, 29.755428197131724),
 (15.199999999999999, 13.753534368139102),
 (21.5, 21.409113198677844),
 (20.0, 16.888073299864597),
 (23.199999999999999, 16.800254503065212),
 (13.800000000000001, 12.568970591325229),
 (24.0, 24.492314764025906),
 (18.300000000000001, 19.310096405422847),
 (23.199999999999999, 21.623648491194352),
 (24.600000000000001, 28.606987897754269),
 (11.9, 7.7733617879571231),
 (28.600000000000001, 28.255550464930067),
 (24.800000000000001, 30.342144644971

In [40]:
y_l_pred=l_obj.predict(X_test)

In [41]:
r2_l_scr=r2_score(y_test,y_l_pred)

In [42]:
r2_l_scr

0.76228794692780422

In [43]:
msqer_l_scr=mean_squared_error(y_test,y_l_pred)

In [44]:
msqer_l_scr

23.17870908345494

In [45]:
print(l_obj.coef_)

[-0.63884494  0.74860597 -0.          0.76800863 -1.57253989  2.5914499  -0.
 -2.46229842  1.38591055 -1.05764651 -1.83668022  0.76597462 -3.83776427]


Here, alpha is a regularisation parameter. For the Ridge model the value of alpha is considered 0.001 and for Lasso Model is 0.1

In [ ]:
R^2 Value for Lasso :0.76228794692780422
MSE Value for Lasso :23.17870908345494